In [ ]:
import requests
import pandas as pd
import os
import urllib.request
from datetime import timedelta

### Lista de ativos

In [ ]:
caminho_dados= '/home/brenno/Documentos/dados/videos_youtube/2024/janeiro'
chave_api = 'fFsjV6Q89fCVehoFPa1o1z9YrWsskP12nKFAhjO4'
headers = {'accept': 'application/json',
                'X-API-Key': chave_api}
os.chdir(caminho_dados)

In [ ]:
response = requests.get(f'https://api.fintz.com.br/bolsa/b3/avista/busca',  
                                        headers=headers)
df = pd.DataFrame(response.json())
df

In [39]:
df[df['ticker'] == 'LWSA3']

ticker  nome  ativo
1645  LWSA3  LWSA   True

In [ ]:
df[df['nome'].str.contains('RAIA')]

In [32]:
cotacoes = pd.read_parquet("/home/brenno/Documentos/codigos_curso/galaxia_11_factor_investing/dados/cotacoes.parquet")
cotacoes['data'] = pd.to_datetime(cotacoes['data']).dt.date
cotacoes['ticker'] = cotacoes['ticker'].astype(str)
cotacoes = cotacoes.sort_values('data', ascending=True)
cotacoes

data  preco_abertura  preco_fechamento  preco_maximo  \
0        2010-01-04         6.70000            6.8000        6.8000   
230      2010-01-04         7.31000            7.3300        7.3300   
229      2010-01-04        26.00000           25.6300       26.0900   
228      2010-01-04        16.29000           16.4000       16.4700   
227      2010-01-04         0.00028            0.0003        0.0003   
...             ...             ...               ...           ...   
1666789  2024-01-12        11.81000           11.9000       11.9800   
1666788  2024-01-12        12.45000           12.4300       12.6700   
1666787  2024-01-12        11.42000           11.8500       11.8800   
1666796  2024-01-12         0.64000            0.6400        0.6500   
1667039  2024-01-12         9.92000            9.8300       10.1900   

         preco_medio  preco_minimo  quantidade_negociada  quantidade_negocios  \
0            6.74000       6.58000                514700                  849   
230          7.31000       7.31000                   500                    3   
229         25.73000      25.50000               1037000                 2528   
228         16.34000      16.03000                182400                  556   
227          0.00029       0.00028              37200000                   16   
...              ...           ...                   ...                  ...   
1666789     11.79000      11.40000                366800                 1265   
1666788     12.53000      12.43000               1555800                 5137   
1666787     11.76000      11.42000                   700                    7   
1666796      0.63000       0.63000               2682300                  462   
1667039      9.86000       9.69000                563400                 2536   

        ticker  volume_negociado  fator_ajuste  preco_fechamento_ajustado  \
0        PMAM3         3469637.0      15.49440                     105.36   
230      TRPN3            3659.0       0.07639                       0.56   
229      LIGT3        26676044.0       0.54253                      13.91   
228      SLCE3         2980598.0       0.12524                       2.05   
227      STRP4           10947.0       1.00000                       0.00   
...        ...               ...           ...                        ...   
1666789  MDNE3         4326337.0       1.00000                      11.90   
1666788  AESB3        19496523.0       1.00000                      12.43   
1666787  PTNT3            8237.0       1.00000                      11.85   
1666796  OIBR3         1714713.0       1.00000                       0.64   
1667039  SRNA3         5555131.0       1.00000                       9.83   

         preco_abertura_ajustado  preco_maximo_ajustado  preco_medio_ajustado  \
0                     103.812480             105.361920            104.432256   
230                     0.558411               0.559939              0.558411   
229                    14.105780              14.154608             13.959297   
228                     2.040160               2.062703              2.046422   
227                     0.000280               0.000300              0.000290   
...                          ...                    ...                   ...   
1666789                11.810000              11.980000             11.790000   
1666788                12.450000              12.670000             12.530000   
1666787                11.420000              11.880000             11.760000   
1666796                 0.640000               0.650000              0.630000   
1667039                 9.920000              10.190000              9.860000   

         preco_minimo_ajustado  
0                   101.953152  
230                   0.558411  
229                  13.834515  
228                   2.007597  
227                   0.000280  
...                        ...  
1666789              11.400000  
1666788              12.430

In [33]:
cotacoes_com_nomes = pd.merge(df, cotacoes, on = 'ticker')

In [ ]:
cotacoes_com_nomes[cotacoes_com_nomes['ticker'] == 'LWSA3']

In [34]:
ipos = pd.read_excel("IPOS.xlsx")
ipos = ipos[ipos['CLASSIFICAÇÃO'] == 'IPO']
ipos = ipos[ipos['INÍCIO DE NEGOCIAÇÃO'] > "2010-01-01"]
ipos = ipos[~ipos['LISTAGEM NA OFERTA'].str.contains('BDR')]
ipos = ipos[['NOME DE PREGÃO', 'INÍCIO DE NEGOCIAÇÃO', 'PREÇO POR AÇÃO / UNIT / BDR (LÇTO.)4']]
ipos.columns = ['nome', 'data', 'preco']
ipos['nome'] = ipos['nome'].str.replace("¹", "")
ipos['nome'] = ipos['nome'].str.replace("²", "")
ipos

nome       data   preco
188      ALIANSCE 2010-01-29    9.00
190     MULTIPLUS 2010-02-05   16.00
192    BR PROPERT 2010-03-08   13.00
193    OSX BRASIL 2010-03-22  800.00
195   ECORODOVIAS 2010-04-01    9.50
..            ...        ...     ...
464         VIVEO 2021-08-09   19.92
465        RAIZEN 2021-08-05    7.40
466  ONCOCLINICAS 2021-08-10   19.75
467    KORA SAUDE 2021-08-13    7.20
468        VITTIA 2021-09-02    8.60

[126 rows x 3 columns]

In [37]:
cotacoes_com_nomes.nome.sort_values().unique()

array(['3R PETROLEUM', '3TENTOS', 'ABC BRASIL', 'ABNOTE', 'ABRIL EDUCA',
       'ABYARA', 'ACO ALTONA', 'ACOS VILL', 'ADVANCED-DH', 'AERIS',
       'AES BRASIL', 'AES ELPA', 'AES TIETE', 'AES TIETE E', 'AFLUENTE',
       'AFLUENTE T', 'AGCONCESSOES', 'AGRA INCORP', 'AGRE EMP IMO',
       'AGROGALAXY', 'ALFA CONSORC', 'ALFA FINANC', 'ALFA HOLDING',
       'ALFA INVEST', 'ALIANSCE', 'ALIANSCSONAE', 'ALIPERTI',
       'ALL AMER LAT', 'ALL ORE', 'ALLIAR', 'ALLIED', 'ALLOS',
       'ALPARGATAS', 'ALPER S.A.', 'ALPHAVILLE', 'ALUPAR', 'AMAZONIA',
       'AMBEV', 'AMBEV S/A', 'AMBIPAR', 'AMERICANAS', 'AMERICEL', 'AMIL',
       'AMPLA ENERG', 'AMPLA INVEST', 'ANHANGUERA', 'ANIMA', 'AREZZO CO',
       'ARMAC', 'ARTERIS', 'ASSAI', 'ATMASA', 'ATOMPAR', 'AUREN',
       'AUTOMETAL', 'AZEVEDO', 'AZUL', 'B TECH EQI', 'B2W DIGITAL', 'B3',
       'BAHEMA', 'BANCO BMG', 'BANCO INTER', 'BANCO PAN', 'BANESE',
       'BANESTES', 'BANPARA', 'BANRISUL', 'BARDELLA', 'BATTISTELLA',
       'BAUMER', 'BBSEGURIDAD

In [ ]:
ipos[ipos['nome'] == 'LOCAWEB']

In [ ]:
lista_ipos = ipos['nome'].to_list()

In [38]:
lista_ipos[0]

'ALIANSCE'

In [ ]:
cotacoes_com_nomes

In [ ]:
cotacoes_com_nomes = cotacoes_com_nomes[cotacoes_com_nomes['nome'].isin(lista_ipos)]
cotacoes_com_nomes = cotacoes_com_nomes[cotacoes_com_nomes['nome'] != 'HRT PETROLEO']

In [ ]:
cotacoes_com_nomes[cotacoes_com_nomes['ticker'] == 'LSWA3']
cotacoes_com_nomes

In [ ]:
df_retornos_ipos = pd.DataFrame(columns = ['retorno_12m', 'retorno_24m', 'retorno_36m',
                                          'retorno_ibov_12m', 'retorno_ibov_24m', 'retorno_ibov_36m'], 
                                            index = lista_ipos)

In [ ]:
ibovespa = pd.read_parquet("ibov.parquet")
ibovespa['data'] = pd.to_datetime(ibovespa['data']).dt.date

In [ ]:
for empresa in lista_ipos:
    
    preco_de_ipo = ipos[ipos['nome'] == empresa]['preco'].iloc[0]
    cotacoes_empresa = cotacoes_com_nomes[cotacoes_com_nomes['nome'] == empresa]
    if cotacoes_empresa.empty == False:
        fator_de_ajuste = cotacoes_empresa['fator_ajuste'].iloc[0]
        preco_ipo_ajustado = preco_de_ipo * fator_de_ajuste
        data_inicial = cotacoes_empresa['data'].iloc[0]
        meses_12 = data_inicial + timedelta(days = 365)
        meses_24 = data_inicial + timedelta(days = 365 * 2)
        meses_36 = data_inicial + timedelta(days = 365 * 3)

        preco_12_meses = cotacoes_empresa[cotacoes_empresa['data'] < meses_12]['preco_fechamento_ajustado'].iloc[-1]
        preco_24_meses = cotacoes_empresa[cotacoes_empresa['data'] < meses_24]['preco_fechamento_ajustado'].iloc[-1]
        preco_36_meses = cotacoes_empresa[cotacoes_empresa['data'] < meses_36]['preco_fechamento_ajustado'].iloc[-1]

        retorno_12m = preco_12_meses/preco_ipo_ajustado - 1
        retorno_24m = preco_24_meses/preco_ipo_ajustado - 1
        retorno_36m = preco_36_meses/preco_ipo_ajustado - 1

        ibovespa_12m = ibovespa[(ibovespa['data'] >= data_inicial) & (ibovespa['data'] < meses_12)]
        ibovespa_24m = ibovespa[(ibovespa['data'] >= data_inicial) & (ibovespa['data'] < meses_24)]
        ibovespa_36m = ibovespa[(ibovespa['data'] >= data_inicial) & (ibovespa['data'] < meses_36)]

        retorno_ibov_12m = ibovespa_12m['fechamento'].iloc[-1]/ibovespa_12m['fechamento'].iloc[0] - 1
        retorno_ibov_24m = ibovespa_24m['fechamento'].iloc[-1]/ibovespa_24m['fechamento'].iloc[0] - 1
        retorno_ibov_36m = ibovespa_36m['fechamento'].iloc[-1]/ibovespa_36m['fechamento'].iloc[0] - 1


        df_retornos_ipos.loc[empresa] = [retorno_12m, retorno_24m, retorno_36m, 
                                         retorno_ibov_12m, retorno_ibov_24m, retorno_ibov_36m]  

In [ ]:
df_retornos_ipos.loc[df_retornos_ipos['retorno_24m'] == df_retornos_ipos['retorno_12m'], 'retorno_24m'] = pd.NA
df_retornos_ipos.loc[df_retornos_ipos['retorno_36m'] == df_retornos_ipos['retorno_24m'], 'retorno_36m'] = pd.NA
df_retornos_ipos.loc[df_retornos_ipos['retorno_36m'] == df_retornos_ipos['retorno_12m'], 'retorno_36m'] = pd.NA

In [ ]:
df_retornos_ipos

In [ ]:
df_retornos_ipos['maior_12'] = df_retornos_ipos['retorno_12m'] > df_retornos_ipos['retorno_ibov_12m']
df_retornos_ipos['maior_24'] = df_retornos_ipos['retorno_24m'] > df_retornos_ipos['retorno_ibov_24m']
df_retornos_ipos['maior_36'] = df_retornos_ipos['retorno_36m'] > df_retornos_ipos['retorno_ibov_36m']

df_retorno_ibov = df_retornos_ipos[['maior_12', 'maior_24', 'maior_36']]

In [ ]:
df_retornos_ipos.isna().sum()

# Análise de retornos 

In [ ]:
tabela_retornos_positivos = pd.DataFrame(columns = ['% da amostra'], index = ['12 meses', '24 meses', '36 meses'])
tabela_retornos_acima_do_ibov = pd.DataFrame(columns = ['% da amostra'], index = ['12 meses', '24 meses', '36 meses'])
empresas_com_maiores_retornos = pd.DataFrame(columns = ['Retorno', 'Empresa'], index = ['12 meses', '24 meses', '36 meses'])
empresas_com_menores_retornos = pd.DataFrame(columns = ['Retorno', 'Empresa'], index = ['12 meses', '24 meses', '36 meses'])
medias_e_medianas_de_retorno = pd.DataFrame(columns = ['Média', 'Mediana'], index = ['12 meses', '24 meses', '36 meses'])
composição_da_oferta = pd.DataFrame(columns = ['Varejo', 'Institucional', 'Estrangeiros', 'Outros'], 
                                   index = ['12M > 0', '12M < 0', '24M > 0', '24M < 0', '36M > 0', '36M < 0'])

In [ ]:
(df_retorno_ibov.sum()/len(df_retorno_ibov)).iloc[1]

In [ ]:
tabela_retornos_acima_do_ibov.iloc[0, 0] = str(round((df_retorno_ibov.sum()/len(df_retorno_ibov)).iloc[0] * 100, 2)) + "%"
tabela_retornos_acima_do_ibov.iloc[1, 0] = str(round((df_retorno_ibov.sum()/len(df_retorno_ibov)).iloc[1] * 100, 2)) + "%"
tabela_retornos_acima_do_ibov.iloc[2, 0] = str(round((df_retorno_ibov.sum()/len(df_retorno_ibov)).iloc[2] * 100, 2)) + "%"

In [ ]:
meses = [12, 24, 36]

for i, mes in enumerate(meses):

    df_retornos_ipos_xm = df_retornos_ipos[f'retorno_{mes}m']
    df_retornos_ipos_xm.dropna(inplace = True)
    df_retornos_ipos_xm = df_retornos_ipos_xm.astype(float)
    total_positivo = round((df_retornos_ipos_xm > 0).sum()/len(df_retornos_ipos_xm) * 100, 2) 
    
    tabela_retornos_positivos.iloc[i, 0] = str(total_positivo) + '%'
    
    retorno = round(df_retornos_ipos_xm.max() * 100, 2)
    empresa = df_retornos_ipos_xm.idxmax()
    
    empresas_com_maiores_retornos.iloc[i, 0] = retorno
    empresas_com_maiores_retornos.iloc[i, 1] = empresa

    retorno = round(df_retornos_ipos_xm.min() * 100, 2)
    empresa = df_retornos_ipos_xm.idxmin()
    
    empresas_com_menores_retornos.iloc[i, 0] = retorno
    empresas_com_menores_retornos.iloc[i, 1] = empresa

In [ ]:
medias_e_medianas_de_retorno.iloc[0, 0] = str(round((df_retornos_ipos.mean().iloc[0] * 100), 2)) + "%"
medias_e_medianas_de_retorno.iloc[0, 1] = str(round((df_retornos_ipos.median().iloc[0] * 100), 2)) + "%"
medias_e_medianas_de_retorno.iloc[1, 0] = str(round((df_retornos_ipos.mean().iloc[1] * 100), 2)) + "%"
medias_e_medianas_de_retorno.iloc[1, 1] = str(round((df_retornos_ipos.median().iloc[1] * 100), 2)) + "%"
medias_e_medianas_de_retorno.iloc[2, 0] = str(round((df_retornos_ipos.mean().iloc[2] * 100), 2)) + "%"
medias_e_medianas_de_retorno.iloc[2, 1] = str(round((df_retornos_ipos.median().iloc[2] * 100), 2)) + "%"

display(medias_e_medianas_de_retorno)

In [ ]:
ipos_2 = pd.read_excel("IPOS.xlsx")
ipos_2 = ipos_2[ipos_2['CLASSIFICAÇÃO'] == 'IPO']
ipos_2 = ipos_2[ipos_2['INÍCIO DE NEGOCIAÇÃO'] > "2010-01-01"]
ipos_2 = ipos_2[~ipos_2['LISTAGEM NA OFERTA'].str.contains('BDR')]
ipos_2

In [ ]:
ofertas_12m_maior_que_zero = df_retornos_ipos[df_retornos_ipos['retorno_12m'] > 0].index
comp_ofertas_12m_maior_que_zero = ipos_2[ipos_2['NOME DE PREGÃO'].isin(ofertas_12m_maior_que_zero)]
comp_ofertas_12m_maior_que_zero = comp_ofertas_12m_maior_que_zero[['VAREJO', 'INSTITUCIONAL', 
                                                                   'ESTRANGEIROS', 'OUTROS']]
composição_da_oferta.loc['12M > 0'] = [str(x) + "%" for x in round(comp_ofertas_12m_maior_que_zero.mean() * 100, 2).to_list()]

ofertas_12m_maior_que_zero = df_retornos_ipos[df_retornos_ipos['retorno_12m'] < 0].index
comp_ofertas_12m_maior_que_zero = ipos_2[ipos_2['NOME DE PREGÃO'].isin(ofertas_12m_maior_que_zero)]
comp_ofertas_12m_maior_que_zero = comp_ofertas_12m_maior_que_zero[['VAREJO', 'INSTITUCIONAL', 
                                                                   'ESTRANGEIROS', 'OUTROS']]
composição_da_oferta.loc['12M < 0'] = [str(x) + "%" for x in round(comp_ofertas_12m_maior_que_zero.mean() * 100, 2).to_list()]

ofertas_12m_maior_que_zero = df_retornos_ipos[df_retornos_ipos['retorno_24m'] > 0].index
comp_ofertas_12m_maior_que_zero = ipos_2[ipos_2['NOME DE PREGÃO'].isin(ofertas_12m_maior_que_zero)]
comp_ofertas_12m_maior_que_zero = comp_ofertas_12m_maior_que_zero[['VAREJO', 'INSTITUCIONAL', 
                                                                   'ESTRANGEIROS', 'OUTROS']]
composição_da_oferta.loc['24M > 0'] = [str(x) + "%" for x in round(comp_ofertas_12m_maior_que_zero.mean() * 100, 2).to_list()]

ofertas_12m_maior_que_zero = df_retornos_ipos[df_retornos_ipos['retorno_24m'] < 0].index
comp_ofertas_12m_maior_que_zero = ipos_2[ipos_2['NOME DE PREGÃO'].isin(ofertas_12m_maior_que_zero)]
comp_ofertas_12m_maior_que_zero = comp_ofertas_12m_maior_que_zero[['VAREJO', 'INSTITUCIONAL', 
                                                                   'ESTRANGEIROS', 'OUTROS']]
composição_da_oferta.loc['24M < 0'] = [str(x) + "%" for x in round(comp_ofertas_12m_maior_que_zero.mean() * 100, 2).to_list()]

ofertas_12m_maior_que_zero = df_retornos_ipos[df_retornos_ipos['retorno_36m'] > 0].index
comp_ofertas_12m_maior_que_zero = ipos_2[ipos_2['NOME DE PREGÃO'].isin(ofertas_12m_maior_que_zero)]
comp_ofertas_12m_maior_que_zero = comp_ofertas_12m_maior_que_zero[['VAREJO', 'INSTITUCIONAL', 
                                                                   'ESTRANGEIROS', 'OUTROS']]
composição_da_oferta.loc['36M > 0'] = [str(x) + "%" for x in round(comp_ofertas_12m_maior_que_zero.mean() * 100, 2).to_list()]

ofertas_12m_maior_que_zero = df_retornos_ipos[df_retornos_ipos['retorno_36m'] < 0].index
comp_ofertas_12m_maior_que_zero = ipos_2[ipos_2['NOME DE PREGÃO'].isin(ofertas_12m_maior_que_zero)]
comp_ofertas_12m_maior_que_zero = comp_ofertas_12m_maior_que_zero[['VAREJO', 'INSTITUCIONAL', 
                                                                   'ESTRANGEIROS', 'OUTROS']]
composição_da_oferta.loc['36M < 0'] = [str(x) + "%" for x in round(comp_ofertas_12m_maior_que_zero.mean() * 100, 2).to_list()]

composição_da_oferta

# Relatório de IPOs

### 126 empresas fizeram IPO desde 2010.

<br>
Das 126 empresas:

* 12 não apresentaram dados de 12 meses após IPO.
* 15 não apresentaram dados de 24 meses após IPO.
* 21 não apresentaram após de 36 meses após IPO.

# Fonte dos dados:

- Base de dados da Varos
- B3

### Investindo em evidências

## % das empresas com retorno positivo

In [ ]:
tabela_retornos_positivos

## % das empresas que superaram o ibovespa:

In [ ]:
tabela_retornos_acima_do_ibov

## Empresas com maiores e menores retornos

In [ ]:
display(empresas_com_maiores_retornos)
empresas_com_menores_retornos

## Médias e medianas de retorno

In [ ]:
medias_e_medianas_de_retorno

## Composição das ofertas

In [ ]:
composição_da_oferta